client/agent-opentracing/src/main/java/org/hawkular/apm/agent/opentracing/OpenTracingManager.java

```diff
@@ -0,0 +1,441 @@
+            }
+        }
+    }
+
+    /**
+     * This method attempts to resume a previously suspended trace state
+     * associated with the supplied id. When resumed, the trace state is
+     * associated with the current thread.
+     *
+     * @param id The id of the trace state to resume
+     */
+    public void resume(String id) {
+        try {
+            suspendedStateLock.lock();
+
+            TraceState ts = suspendedState.get(id);
+
+            if (ts != null) {
+                clearExpire(ts);
+
+                // Log after finding trace state, otherwise may generate alot of logging
+                if (log.isLoggable(Level.FINEST)) {
+                    log.finest("Resume trace state = " + ts + " id = " + id);
+                }
+
+                // Check if thread already used
+                if (traceState.get() != null && log.isLoggable(Level.FINEST)) {
+                    log.finest("WARNING: Overwriting previous trace state = " + traceState.get());
+                }
+
+                traceState.set(ts);
+
+                suspendedState.remove(id);
+            }
+        } finally {
+            suspendedStateLock.unlock();
+        }
+    }
+
+    private static void setExpire(TraceState ts) {
+        // If too inefficient getting timestamp each time, could have a set
+        // of used trace states, and when schedule does its cycle, it could
+        // assign a single timestamp to all the trace states that have been
+        // accessed during that cycle
+        ts.setExpire(System.currentTimeMillis() + expiryInterval);
+    }
+
+    private static void clearExpire(TraceState ts) {
+        ts.setExpire(0);
+    }
+
+    private static void cleanup() {
+        try {
+            suspendedStateLock.lock();
+
+            long timestamp = System.currentTimeMillis();
+            Iterator<Map.Entry<String, TraceState>> iter = suspendedState.entrySet().iterator();
+
+            while (iter.hasNext()) {
+                Map.Entry<String, TraceState> entry = iter.next();
+                TraceState ts = entry.getValue();
+                if (ts.getExpire() > 0 && ts.getExpire() < timestamp) {
+                    if (log.isLoggable(Level.FINEST)) {
+                        log.finest("Expired trace state = " + ts + " id = " + entry.getKey());
+                    }
+                    iter.remove();
+                }
+            }
+        } finally {
+            suspendedStateLock.unlock();
+        }
+    }
+
+    /**
+     * This method resets the current thread, and suspended trace states, for testing
+     * purposes.
+     */
+    public static void reset() {
+        traceState.remove();
+        suspendedState.clear();
+    }
+
+    /**
+     * This class represents the state information being accumulated for a
+     * trace instance.
+     *
+     * @author gbrown
+     */
+    public static class TraceState {
+
+        private Stack<Span> spanStack = new Stack<>();
```

Comment:
I'm asking because `Stack` is thread safe. Javadoc advise to use `Deque` in preference to this class.